packages/flutter/lib/src/material/ink_well.dart

```diff
@@ -2,331 +2,193 @@
-}
+class _InkResponseState<T extends InkResponse> extends State<T> {
 
+  bool get containedInWell => false;
 
-double _getSplashTargetSize(Size bounds, Point position) {
-  double d1 = (position - bounds.topLeft(Point.origin)).distance;
-  double d2 = (position - bounds.topRight(Point.origin)).distance;
-  double d3 = (position - bounds.bottomLeft(Point.origin)).distance;
-  double d4 = (position - bounds.bottomRight(Point.origin)).distance;
-  return math.max(math.max(d1, d2), math.max(d3, d4)).ceil().toDouble();
-}
+  Set<InkSplash> _splashes;
+  InkSplash _currentSplash;
 
-class _InkSplash {
-  _InkSplash(this.position, this.renderer) {
-    _targetRadius = _getSplashTargetSize(renderer.size, position);
-    _radius = new ValuePerformance<double>(
-      variable: new AnimatedValue<double>(
-        _kSplashInitialSize,
-        end: _targetRadius,
-        curve: Curves.easeOut
-      ),
-      duration: new Duration(milliseconds: (_targetRadius / _kSplashUnconfirmedVelocity).floor())
-    )..addListener(_handleRadiusChange)
-     ..play();
+  void _handleTapDown(Point position) {
+    RenderBox referenceBox = context.findRenderObject();
+    assert(Material.of(context) != null);
+    InkSplash splash;
+    splash = Material.of(context).splashAt(
+      referenceBox: referenceBox,
+      position: referenceBox.globalToLocal(position),
+      containedInWell: containedInWell,
+      onRemoved: () {
+        if (_splashes != null) {
+          assert(_splashes.contains(splash));
+          _splashes.remove(splash);
+          if (_currentSplash == splash)
+            _currentSplash = null;
+        } // else we're probably in deactivate()
+      }
+    );
+    _splashes ??= new Set<InkSplash>();
+    _splashes.add(splash);
+    _currentSplash = splash;
   }
 
-  final Point position;
-  final _RenderInkSplashes renderer;
-
-  double _targetRadius;
-  double _pinnedRadius;
-  ValuePerformance<double> _radius;
-
-  void _updateVelocity(double velocity) {
-    int duration = (_targetRadius / velocity).floor();
-    _radius.duration = new Duration(milliseconds: duration);
-    _radius.play();
+  void _handleTap() {
+    _currentSplash?.confirm();
+    _currentSplash = null;
+    if (config.onTap != null)
+      config.onTap();
   }
 
-  void confirm() {
-    _updateVelocity(_kSplashConfirmedVelocity);
-    _pinnedRadius = null;
+  void _handleTapCancel() {
+    _currentSplash?.cancel();
+    _currentSplash = null;
   }
 
-  void cancel() {
-    _updateVelocity(_kSplashCanceledVelocity);
-    _pinnedRadius = _radius.value;
+  void _handleDoubleTap() {
+    _currentSplash?.confirm();
+    _currentSplash = null;
+    if (config.onDoubleTap != null)
+      config.onDoubleTap();
   }
 
-  void _handleRadiusChange() {
-    if (_radius.value == _targetRadius)
-      renderer._removeSplash(this);
-    renderer.markNeedsPaint();
+  void _handleLongPress() {
+    _currentSplash?.confirm();
+    _currentSplash = null;
+    if (config.onLongPress != null)
```

Comment:
I don't think it's possible for this to be null here, but it seems find to handle that case for symmetry.